commit 0002f1f94ccce9fbf8241ad26e3bcdeb360d82dc
Author: randymcmillan <randymcmillan@protonmail.com>
Date:   Tue Dec 2 12:57:59 2025 -0500

    feat(types): Update NIP-11 RelayInformationDocument
    
    00-00000f4e

diff --git a/src/lib/types/versioned/relay_information_document2.rs b/src/lib/types/versioned/relay_information_document2.rs
index 6cbef50c82..63fc3b6f67 100644
--- a/src/lib/types/versioned/relay_information_document2.rs
+++ b/src/lib/types/versioned/relay_information_document2.rs
@@ -126,7 +126,7 @@ impl fmt::Display for RelayLimitationV2 {
 }
 
 /// Relay information document as described in NIP-11, supplied by a relay
-#[derive(Clone, Debug, PartialEq, Eq)]
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
 pub struct RelayInformationDocumentV2 {
     /// Name of the relay
     pub name: Option<String>,
@@ -134,9 +134,24 @@ pub struct RelayInformationDocumentV2 {
     /// Description of the relay in plain text
     pub description: Option<String>,
 
+    /// A banner image for the relay
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    pub banner: Option<Url>,
+
+    /// An icon for the relay
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    pub icon: Option<Url>,
+
     /// Public key of an administrative contact of the relay
     pub pubkey: Option<PublicKeyHex>,
 
+    /// The relay's public key, for signing events
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[serde(default)]
+    pub self_pubkey: Option<PublicKeyHex>,
+
     /// An administrative contact for the relay. Should be a URI.
     pub contact: Option<String>,
 
@@ -182,7 +197,10 @@ impl Default for RelayInformationDocumentV2 {
         RelayInformationDocumentV2 {
             name: None,
             description: None,
+            banner: None,
+            icon: None,
             pubkey: None,
+            self_pubkey: None,
             contact: None,
             supported_nips: vec![],
             software: None,
@@ -220,7 +238,10 @@ impl RelayInformationDocumentV2 {
         RelayInformationDocumentV2 {
             name: Some("Crazy Horse".to_string()),
             description: Some("A really wild horse".to_string()),
+            banner: Some(Url::try_from_str("https://example.com/banner.jpg").unwrap()),
+            icon: Some(Url::try_from_str("https://example.com/icon.jpg").unwrap()),
             pubkey: Some(PublicKeyHex::mock()),
+            self_pubkey: Some(PublicKeyHex::mock()),
             contact: None,
             supported_nips: vec![11, 12, 13, 14],
             software: None,
@@ -317,9 +338,18 @@ impl fmt::Display for RelayInformationDocumentV2 {
         if let Some(desc) = &self.description {
             write!(f, " Description=\"{desc}\"")?;
         }
+        if let Some(banner) = &self.banner {
+            write!(f, " Banner=\"{banner}\"")?;
+        }
+        if let Some(icon) = &self.icon {
+            write!(f, " Icon=\"{icon}\"")?;
+        }
         if let Some(pubkey) = &self.pubkey {
             write!(f, " Pubkey=\"{pubkey}\"")?;
         }
+        if let Some(self_pubkey) = &self.self_pubkey {
+            write!(f, " SelfPubkey=\"{self_pubkey}\"")?;
+        }
         if let Some(contact) = &self.contact {
             write!(f, " Contact=\"{contact}\"")?;
         }
@@ -375,140 +405,7 @@ impl fmt::Display for RelayInformationDocumentV2 {
     }
 }
 
-impl Serialize for RelayInformationDocumentV2 {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        let mut map = serializer.serialize_map(Some(7 + self.other.len()))?;
-        if self.name.is_some() {
-            map.serialize_entry("name", &json!(&self.name))?;
-        }
-        if self.description.is_some() {
-            map.serialize_entry("description", &json!(&self.description))?;
-        }
-        if self.pubkey.is_some() {
-            map.serialize_entry("pubkey", &json!(&self.pubkey))?;
-        }
-        if self.contact.is_some() {
-            map.serialize_entry("contact", &json!(&self.contact))?;
-        }
-        map.serialize_entry("supported_nips", &json!(&self.supported_nips))?;
-        if self.software.is_some() {
-            map.serialize_entry("software", &json!(&self.software))?;
-        }
-        if self.version.is_some() {
-            map.serialize_entry("version", &json!(&self.version))?;
-        }
-        if self.limitation.is_some() {
-            map.serialize_entry("limitation", &json!(&self.limitation))?;
-        }
-        if !self.retention.is_empty() {
-            map.serialize_entry("retention", &json!(&self.retention))?;
-        }
-        if !self.relay_countries.is_empty() {
-            map.serialize_entry("relay_countries", &json!(&self.relay_countries))?;
-        }
-        if !self.language_tags.is_empty() {
-            map.serialize_entry("language_tags", &json!(&self.language_tags))?;
-        }
-        if !self.tags.is_empty() {
-            map.serialize_entry("tags", &json!(&self.tags))?;
-        }
-        if self.posting_policy.is_some() {
-            map.serialize_entry("posting_policy", &json!(&self.posting_policy))?;
-        }
-        if self.payments_url.is_some() {
-            map.serialize_entry("payments_url", &json!(&self.payments_url))?;
-        }
-        if self.fees.is_some() {
-            map.serialize_entry("fees", &json!(&self.fees))?;
-        }
-        for (k, v) in &self.other {
-            map.serialize_entry(&k, &v)?;
-        }
-        map.end()
-    }
-}
-
-impl<'de> Deserialize<'de> for RelayInformationDocumentV2 {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        deserializer.deserialize_map(RidVisitor)
-    }
-}
-
-struct RidVisitor;
-
-impl<'de> Visitor<'de> for RidVisitor {
-    type Value = RelayInformationDocumentV2;
 
-    fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "A JSON object")
-    }
-
-    fn visit_map<M>(self, mut access: M) -> Result<RelayInformationDocumentV2, M::Error>
-    where
-        M: MapAccess<'de>,
-    {
-        let mut map: Map<String, Value> = Map::new();
-        while let Some((key, value)) = access.next_entry::<String, Value>()? {
-            let _ = map.insert(key, value);
-        }
-
-        let mut rid: RelayInformationDocumentV2 = Default::default();
-
-        if let Some(Value::String(s)) = map.remove("name") {
-            rid.name = Some(s);
-        }
-        if let Some(Value::String(s)) = map.remove("description") {
-            rid.description = Some(s);
-        }
-        if let Some(Value::String(s)) = map.remove("pubkey") {
-            rid.pubkey = PublicKeyHex::try_from_string(s).ok();
-        }
-        if let Some(Value::String(s)) = map.remove("contact") {
-            rid.contact = Some(s);
-        }
-        if let Some(Value::Array(vec)) = map.remove("supported_nips") {
-            for elem in vec.iter() {
-                if let Value::Number(num) = elem {
-                    if let Some(u) = num.as_u64() {
-                        rid.supported_nips.push(u as u32);
-                    }
-                }
-            }
-        }
-        if let Some(Value::String(s)) = map.remove("software") {
-            rid.software = Some(s);
-        }
-        if let Some(Value::String(s)) = map.remove("version") {
-            rid.version = Some(s);
-        }
-        if let Some(v) = map.remove("limitation") {
-                        rid.limitation = serde_json::from_value::<Option<RelayLimitationV2>>(v).unwrap_or_default()        }
-        if let Some(v) = map.remove("retention") {
-                          rid.retention = serde_json::from_value::<Vec<RelayRetentionV1>>(v).unwrap_or_default();        }
-        if let Some(v) = map.remove("relay_countries") {
-                          rid.relay_countries = serde_json::from_value::<Vec<String>>(v).unwrap_or_default()        }
-        if let Some(v) = map.remove("language_tags") {
-                          rid.language_tags = serde_json::from_value::<Vec<String>>(v).unwrap_or_default()        }
-        if let Some(v) = map.remove("tags") {
-                          rid.tags = serde_json::from_value::<Vec<String>>(v).unwrap_or_default()        }
-        if let Some(v) = map.remove("posting_policy") {
-                          rid.posting_policy = serde_json::from_value::<Option<Url>>(v).unwrap_or_default()        }
-        if let Some(v) = map.remove("payments_url") {
-                          rid.payments_url = serde_json::from_value::<Option<Url>>(v).unwrap_or_default()        }
-        if let Some(v) = map.remove("fees") {
-                          rid.fees = serde_json::from_value::<Option<RelayFeesV1>>(v).unwrap_or_default()        }
-
-        rid.other = map;
-
-        Ok(rid)
-    }
-}
 
 #[cfg(test)]
 mod test {
